src/client/pythonEnvironments/base/info/env.ts

```diff
@@ -0,0 +1,223 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+import { cloneDeep } from 'lodash';
+import * as path from 'path';
+import {
+    FileInfo,
+    PythonDistroInfo,
+    PythonEnvInfo, PythonEnvKind, PythonVersion,
+} from '.';
+import { Architecture } from '../../../common/utils/platform';
+import { arePathsSame } from '../../common/externalDependencies';
+import { areEqualVersions, areEquivalentVersions } from './pythonVersion';
+
+export function areSameEnvironment(
+    left: string | PythonEnvInfo,
+    right: string | PythonEnvInfo,
+    allowPartialMatch?:boolean,
+): boolean {
+    const leftFilename = typeof left === 'string' ? left : left.executable.filename;
+    const rightFilename = typeof right === 'string' ? right : right.executable.filename;
+
+    if (arePathsSame(leftFilename, rightFilename)) {
+        return true;
+    }
+
+    if (arePathsSame(path.dirname(leftFilename), path.dirname(rightFilename))) {
+        const leftVersion = typeof left === 'string' ? undefined : left.version;
+        const rightVersion = typeof right === 'string' ? undefined : right.version;
+        if (leftVersion && rightVersion) {
+            if (
+                areEqualVersions(leftVersion, rightVersion)
+                || (allowPartialMatch && areEquivalentVersions(leftVersion, rightVersion))
+            ) {
+                return true;
+            }
+        }
+    }
+    return false;
+}
+
+/**
+ * Returns a heuristic value on how much information is available in the given version object.
+ * @param {PythonVersion} version version object to generate heuristic from.
+ * @returns A heuristic value indicating the amount of info available in the object
+ * weighted by most important to least important fields.
+ * Wn > Wn-1 + Wn-2 + ... W0
+ */
+function getPythonVersionInfoHeuristic(version:PythonVersion): number {
+    let infoLevel = 0;
+    if (version.major > 0) {
+        infoLevel += 20; // W4
+    }
+
+    if (version.minor >= 0) {
+        infoLevel += 10; // W3
+    }
+
+    if (version.micro >= 0) {
+        infoLevel += 5; // W2
+    }
+
+    if (version.release.level) {
+        infoLevel += 3; // W1
+    }
+
+    if (version.release.serial || version.sysVersion) {
+        infoLevel += 1; // W0
+    }
+
+    return infoLevel;
+}
+
+/**
+ * Returns a heuristic value on how much information is available in the given executable object.
+ * @param {FileInfo} executable executable object to generate heuristic from.
+ * @returns A heuristic value indicating the amount of info available in the object
+ * weighted by most important to least important fields.
+ * Wn > Wn-1 + Wn-2 + ... W0
+ */
+function getFileInfoHeuristic(file:FileInfo): number {
+    let infoLevel = 0;
+    if (file.filename.length > 0) {
+        infoLevel += 5; // W2
+    }
+
+    if (file.mtime) {
+        infoLevel += 2; // W1
+    }
+
+    if (file.ctime || file.mtime) {
+        infoLevel += 1; // W0
+    }
```

Comment:
```suggestion
    if (file.mtime) {
        infoLevel += 3; // W1
    }

    if (file.ctime) {
        infoLevel += 1; // W0
    }
```
`ctime` more important than `mtime`?

Reply:
Oops